Move non application endpoints to separate URL path

kenfinnigan opened the pull request:
- Fixes #13144
- OpenAPI, Metrics, Micrometer, Health, Health UI endpoints are all now served under /q by default
- Previous behavior can be attained by setting `quarkus.http.non-application-root-path` to `/`

Subsequent issue will be created for handling port and binding changes for non application endpoints

kenfinnigan committed:
Move non application endpoints to separate URL path

- Fixes #13144
- OpenAPI, Metrics, Micrometer, Health, Health UI endpoints are all now served under /q by default
- Previous behavior can be attained by setting `quarkus.http.framework-root-path` to `/`

Subsequent issue will be created for handling port and binding changes for non application endpoints

kenfinnigan committed:
Fix some more url references in tests found from CI

cescoffier commented on extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java:
```diff
@@ -154,7 +153,7 @@ MetricsCapabilityBuildItem metricsCapabilityBuildItem() {
     void createRoute(BuildProducer<RouteBuildItem> routes,
             SmallRyeMetricsRecorder recorder,
-            HttpRootPathBuildItem httpRoot,
+            FrameworkRootPathBuildItem frameworkRoot,
```
So, does that mean that Framework is equivalent to NonApplication? 
Can't we have a single word?

kenfinnigan replied:
Yeah, I kind of flip flopped between them.

Leaning towards "NonApplication". If that's preferred I can align them all

cescoffier replied:
Yes, I actually prefer avoiding the word "framework"

gastaldi replied:
How about `Core`?

gastaldi replied:
Um, actually I think something like `management` would make more sense

kenfinnigan replied:
`management` implies you can actually manage something, which isn't the case with these endpoints.

I've changed it to "NonApplication"

cescoffier commented on extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/spring/SwaggerAndOpenAPIWithCommonPrefixTest.java:
```diff
@@ -25,7 +25,7 @@
     @Test
     public void shouldWorkEvenWithCommonPrefix() {
         RestAssured.when().get("/swagger-ui/index.html").then().statusCode(200).body(containsString("/swagger"));
```
Why does swagger-ui stay on the main router? 
I understand it's an asset, but that looks not totally right.

kenfinnigan replied:
My reasoning, for now, is it's a Dev/Test extension only.

Happy to revise either now or in future if that's the preferred approach

cescoffier replied:
it's dev mode only because it was exposed on the _public internet_. But if we fix that "small" issue, it might be very useful even in prod.

kenfinnigan replied:
Fair enough, I can move to "non app" as well

cescoffier commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/FrameworkRootPathBuildItem.java:
```diff
@@ -0,0 +1,33 @@
+    private final boolean separateRoot;
+
+    public FrameworkRootPathBuildItem(String frameworkRootPath) {
+        this.frameworkRootPath = frameworkRootPath;
```
That only configure the "/q".
Don't you want to allow having a separate HTTP server (maybe it's done somewhere else)

kenfinnigan replied:
I've created https://github.com/quarkusio/quarkus/issues/13602 to handle the port and binding options to keep this PR simple and get it in

cescoffier replied:
ah ok!

cescoffier commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java:
```diff
@@ -35,4 +35,13 @@
+     * resulting in all framework endpoints being served from "/" along with the application.
+     */
+    @ConfigItem(defaultValue = "/q")
+    public String frameworkRootPath;
```
So, we have /quarkus for the console and /q for framework.
I may be wrong but I don't believe the console should be "application". 

Also, all guides, workshops and quickstarts would need an update once this is merged

kenfinnigan replied:
Is that the new dev console that was created as part of Reactive Resteasy? Did that get merged already, might have missed it.

If it's in, would /q/console be ok?

cescoffier replied:
I let @FroMage @stuartwdouglas and @geoand decide.

geoand replied:
The console started out as part of RESTEasy Reactive, but is now in a separate (not yet merged) branch. The console currently is under `/@dev`.

FroMage replied:
Who decided on the `/q` default prefix? Isn't it likely to conflict with user paths?

kenfinnigan replied:
It was agreed a few weeks ago in a team meeting, but not set in stone

FroMage replied:
I guess I would have made it `/@q` to make sure it didn't conflict. But if everyone prefers `/q` let's go with that.

As for the dev console it should move to `/q/dev` if we do that.

gastaldi replied:
I think `@` is reserved for a special meaning within a scheme (eg. https://foo:bar@example.com)

gastaldi replied:
But looking at the [URL RFC](https://tools.ietf.org/html/rfc1738) I see that `@` is accepted, so I dunno 😃

FroMage replied:
It's only reserved as part of the `authority` section (host name).

cescoffier commented on tcks/microprofile-health/pom.xml:
```diff
@@ -18,6 +18,10 @@
                 <artifactId>maven-surefire-plugin</artifactId>
                 <configuration>
+                    <systemPropertyVariables>
+                        <!-- Change metrics endpoint to be /metrics -->
```
Should we consider updating the TCKs to allows setting the prefix?

kenfinnigan replied:
Some TCKs explicitly indicate the name of the URL path, sometimes in relation to root context, so I don't think this would be possible

cescoffier commented:
This PR moves the "non-application" endpoints under "/q", but, from what I can see, the endpoints are still exposed on the same server. But nothing is avoiding their access from the regular _route_ (except if you configure a proxy in front that blocks external access to "/q/*").

Also, this is a breaking change, and we would need to document it carefully (as it breaks existing applications).

slaskawi commented:
Thanks for this change, @kenfinnigan !

Is it possible to bind the non-application endpoints to a separate IP address? This way I could bind it to `127.0.0.1`, whereas the application would bind to `0.0.0.0`?

kenfinnigan commented:
@slaskawi separate ports and binding were split into a separate issue (https://github.com/quarkusio/quarkus/issues/13602) to simplify the PR

slaskawi commented:
> @slaskawi separate ports and binding were split into a separate issue (#13602) to simplify the PR

@kenfinnigan Great! That answers my question. Thanks!

kenfinnigan committed:
- Adds documentation updates
- Includes Swagger UI under the new non application endpoint path
- Changes names to refer to "NonApplication" instead of Framework

kenfinnigan committed:
Forgot I fixed this test in GH UI last week

kenfinnigan committed:
Missed a couple of Swagger UI path usages

maxandersen commented:
if we are sure `@` is legal and not reserved in general as I initially interpreted: 
"The characters ";",
   "/", "?", ":", "@", "=" and "&" are the characters which may be
   reserved for special meaning within a scheme"

then `/@q/' or `/@qs/` (qs as in *q*uarku*s*) makes sense to me which is also my suggestion for the cli shorthand.

`@` to avoid obvious overlaps
`qs` to match `quarkus` 

I have same concern as @FroMage that `/q` overlaps with users/customers.

i.e. a common pattern I've spotted in past is that `/q` is used for "query" `/q?param=wefsdf`

kenfinnigan commented:
I'm fine with changing the path to be `/@q`, `/@qs`, `/@qk`, or whatever else people might agree on.

If `@` is confirmed to be valid, that's great.

Given everyone's views differ, how do we reach agreement?

kenfinnigan committed:
Add HTTP Redirect from current path to new non application endpoint root path

kenfinnigan committed:
Minor grammar cleanup in docs

ebullient commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java:
```diff
@@ -117,11 +122,28 @@ public KubernetesPortBuildItem kubernetes() {
+        if (frameworkRouterFound && nonApplicationRootPath.isSeparateRoot()) {
+            // Handle redirects from old paths to new non application endpoint root
+            recorder.addNonApplicationPathRedirect(router, frameworkRouter, nonApplicationRootPath.getFrameworkRootPath());
+
```
Should this be disable-able?

ebullient replied:
I think httpBuildTimeConfig.redirectToNonApplicationRootPath came in while I was reviewing .. ;)

kenfinnigan committed:
- Add ability to disable redirect of non application endpoints to non application root path
- Renamed property for defining non application root path to no longer use "framework"

ebullient commented on integration-tests/main/src/test/java/io/quarkus/it/main/OpenApiTestCase.java:
```diff
@@ -22,7 +22,7 @@
     private static final String DEFAULT_MEDIA_TYPE = "application/json";
 
-    @TestHTTPResource("openapi")
+    @TestHTTPResource("q/openapi")
```
This is one of the only places w/o the leading '/' ... 
while it may work, the inconsistency is noticable

kenfinnigan committed:
Fix my bad substring!

kenfinnigan commented:
@cescoffier when you have a moment could you check on the changes since you last reviewed? Thanks

maxandersen commented:
so we went with `/q` instead of more unique ones ? damn :/
